File: PermutationinString.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if(s1.length() > s2.length()){
            return false;
        }
        int[] s1arr = new int[26];
        int[] s2arr = new int[26];
        for(int i = 0; i < s1.length(); i++){
            s1arr[s1.charAt(i)-'a']++;
            s2arr[s2.charAt(i)-'a']++;
        }
        int left = 0;
        for(int right = s1.length(); right < s2.length(); right++){
            if(checkEqual(s1arr, s2arr)){
                return true;
            }
            s2arr[s2.charAt(right)-'a']++;
            s2arr[s2.charAt(left)-'a']--;
            left++;
        }
        return checkEqual(s1arr, s2arr);
    }
    public boolean checkEqual(int[] s1arr, int[] s2arr){
        for(int i = 0; i < 26; i++){
            if(s1arr[i] != s2arr[i]){
                return false;
            }
        }
        return true;
    }
}

File: PermutationinString.py
class Solution(object):
    def checkInclusion(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        maps1 = {}
        maps2 = {}
        matches = 0
        for i in s1:
            maps1[i] = 1 + maps1.get(i, 0)
        left = 0
        for right in range(len(s2)):
            maps2[s2[right]] = 1 + maps2.get(s2[right], 0)
            if maps2[s2[right]] == maps1.get(s2[right], 0):
                matches += 1
            if matches == len(maps1):
                return True
            if (right-left+1 == len(s1)):
                if maps2[s2[left]] == maps1.get(s2[left], 0):
                    matches -= 1
                maps2[s2[left]] -= 1
                left += 1
        return False


        